### DISCUSSION TITLE
I have a question only for caverns 4

### DISCUSSION THREAD

**Discussion by Raptorq1** [Techokami/SonicWorldsNext#211] [Q&A]
How did you make a working level select its because I am trying to make one

**Comment by r6915ee** [comment]
> How did you make a working level select its because I am trying to make one

It should be most of the same procedure as just making a regular menu that can change the scene, as all levels in Worlds Next are all scenes.

**Reply by Raptorq1** [reply]
I all ready have one all I need help is making it work 

![Screenshot_20241110-005314](https://github.com/user-attachments/assets/383091b8-f9ae-43f5-bbf1-6f899e7b5f28)


Here's the scene tree

 ┖╴CharacterSelect
    ┠╴UI
    ┃  ┠╴Backdrop
    ┃  ┃  ┖╴Back
    ┃  ┠╴Labels
    ┃  ┃  ┖╴LevelsLeft
    ┃  ┃     ┖╴@VScrollBar@2
    ┃  ┖╴LabelsRight
    ┃     ┠╴LevelsRight
    ┃     ┃  ┖╴@VScrollBar@3
    ┃     ┠╴CharacterOrigin
    ┃     ┃  ┠╴Sonic
    ┃     ┃  ┠╴Tails
    ┃     ┃  ┠╴Knuckles
    ┃     ┃  ┠╴Amy
    ┃     ┃  ┠╴Mighty
    ┃     ┃  ┖╴Ray
    ┃     ┠╴Front
    ┃     ┖╴LevelIcon
    ┠╴Cover
    ┃  ┖╴Back
    ┖╴Switch

**Reply by r6915ee** [reply]
The basic way to create a menu of this kind should work. Most of the time, you'd have to implement a choice variable (usually an int), add functionality to refresh the graphics, add the ability to loop between menu choices, and do some action according to the current choice (for example in a match statement for the function).

**Reply by Raptorq1** [reply]
I have already done it  but now I need help making the sound test icon change depending on the character that's selected

[LevelSelect.zip](https://github.com/user-attachments/files/17735100/LevelSelect.zip)

**Reply by r6915ee** [reply]
Once the selected character has changed, you can change the texture or frame to the necessary icon, similarly to how highlighting the text here works.

**Reply by Raptorq1** [reply]
> Once the selected character has changed, you can change the texture or frame to the necessary icon, similarly to how highlighting the text here works.

Oh and how would I Check if tails is visible and if tails is visible it sets it to sonic and tails and if tails is not visible it sets it to sonic alone

**Reply by r6915ee** [reply]
One example is to use a `match` statement to set the characters. `match` statements are essentially GDScript's `switch-case` statements. For example:

```gdscript
match some_var:
    some_value:
        pass
    other_value:
        pass
```

From there, you would use the `match` statement to select the character by checking the current selected character value and then setting the characters in the regular gameplay accordingly.

**Reply by Caverns4** [reply]
You're doing it backwards. Tails being visible or not depends on the character selection, not the other way around.
Look at how the Character Select screen in base Worlds Next, that's what I built my level select off of, just added a bit more functionality.

The only real difference is that the left/right input for the character select is locked behind the menu selection being a specific value (the entire length of the menu -1)

**Reply by Raptorq1** [reply]
> You're doing it backwards. Tails being visible or not depends on the character selection, not the other way around. Look at how the Character Select screen in base Worlds Next, that's what I built my level select off of, just added a bit more functionality.
> 
> The only real difference is that the left/right input for the character select is locked behind the menu selection being a specific value (the entire length of the menu -1)

Okay but the level select that I am trying to recreate is the one in mania and tails becomes visible or invisible if a button is pressed


https://github.com/user-attachments/assets/8b5152b3-6941-41bf-9544-7a4ca0ff6888

**Reply by Raptorq1** [reply]
> You're doing it backwards. Tails being visible or not depends on the character selection, not the other way around. Look at how the Character Select screen in base Worlds Next, that's what I built my level select off of, just added a bit more functionality.
> 
> The only real difference is that the left/right input for the character select is locked behind the menu selection being a specific value (the entire length of the menu -1)

This is the code that handles the tails visible

if characterID ==0 and Input.is_action_just_pressed("gm_super"):
			!$UI/LabelsRight/CharacterOrigin/Tails.visible == $UI/LabelsRight/CharacterOrigin/Tails.visible

**Reply by Raptorq1** [reply]
> You're doing it backwards. Tails being visible or not depends on the character selection, not the other way around. Look at how the Character Select screen in base Worlds Next, that's what I built my level select off of, just added a bit more functionality.
> 
> The only real difference is that the left/right input for the character select is locked behind the menu selection being a specific value (the entire length of the menu -1)

I fixed it

**Comment by Raptorq1** [comment]
I need help I am making a mobile port of my game but I need help checking if its a level that's being selected it switches the input action for the touch control to gm_pause or if its sound test that's being selected it switches to gm_action to allow the music to play

func check():
	match levelIcons.size():
		0, 1:
			$UI/Control/TouchScreenButton.action = "gm_pause"
		28:
			$UI/Control/TouchScreenButton.action = "gm_action"
	print("New action:", $UI/Control/TouchScreenButton.action)

**Comment by DimensionWarped** [comment]
There is literally a level select built into the engine already. A rudimentary one, but a level select never the less. The changes you are wanting to make to it fit into the category of basic interface design and I'll remind you that this forum is not for absolute beginner level Godot questions.
